lib/uri_parser.js

```diff
@@ -296,6 +296,76 @@ function applyConnectionStringOption(obj, key, value, options) {
   obj[key] = value;
 }
 
+const USERNAME_REQUIRED_MECHANISMS = new Set([
+  'GSSAPI',
+  'MONGODB-CR',
+  'PLAIN',
+  'SCRAM-SHA-1',
+  'SCRAM-SHA-256'
+]);
+
+/**
+ * Modifies the parsed connection string object taking into account expectations we
+ * have for authentication-related options.
+ *
+ * @param {object} parsed The parsed connection string result
+ * @return The parsed connection string result possibly modified for auth expectations
+ */
+function applyAuthExpectations(parsed) {
+  if (parsed.options == null) {
+    return;
+  }
+
+  const options = parsed.options;
+  const authSource = options.authsource || options.authSource;
+  if (authSource != null) {
+    parsed.auth = Object.assign({}, parsed.auth, { db: authSource });
+  }
+
+  const authMechanism = options.authmechanism || options.authMechanism;
+  if (authMechanism != null) {
```

Comment:
Can we break up some of the sub-`if`s into sub-functions? This `if` seems really big

Reply:
This is what I was referring to yesterday in our meeting. I think all of these, in a more elegantly designed system, would live in some sort of factory method on each of the sasl modules. As it stands, its a little long but its a literal parse and I'd prefer to actually reduce the function calls unless its bringing a great increase in readability. Do you feel strongly about this?

Reply:
Put this comment on the wrong thread: nah, if we moved it to the authproviders later that would be fine. Can we file a ticket for it?

Reply:
Done. [NODE-1691](https://jira.mongodb.org/browse/NODE-1691)